### PR TITLE
test: tests using fakeredis to check compatibility

### PR DESCRIPTION
**Pull request by cunla**
Fix #3519 

- See example results here: https://cunla.github.io/dragonfly/?sort=result
- Please update workflow to deploy where appropriate (currently it will override existing github pages if exist).
- Notice in the tests code, some tests have `@pytest.mark.unsupported_server_types("dragonfly")  # TODO ...` - These tests should pass in my opinion, but currently cause unexpected behavior

### CODE REVIEW

**Comment by dranikpg** [comment]
Hi! Thanks for your efforts 🎩 

Seems like I'm asking a little late, but wouldn't there be a way to run the tests from your repository with a custom script or dockerfile? See this folder https://github.com/dragonflydb/dragonfly/tree/main/tests/integration for examples where we test node-redis, ioredis, jedis, etc

**Comment by romange** [comment]
@dranikpg I've discussed with @cunla  privately. Seems that these tests will provide a valuable addition to our regression tests and will help with our compatibility efforts.

**Comment by cunla** [comment]
Hi, it is running on my repository (https://github.com/cunla/fakeredis-py/actions/runs/10393435202).

However, on my repository
1) It won't publish to github-pages (as @romange requested).
2) It would run similar to other tests, i.e., less informative than the current report showing every test that failed.
3) Check the information on the disabled tests (`@pytest.mark.unsupported_server_types("dragonfly")  # TODO ...`) - I don't know if you want that in other repos.

File: .github/workflows/test-fakeredis.yml [hunk 1]
```diff
@@ -0,0 +1,152 @@
+---
+name: Test Dragonfly/Fakeredis
+
+on:
```
Comment (romange) [inline .github/workflows/test-fakeredis.yml:4]: seems that you need 
`id-token:write`

directive like here: https://github.com/dragonflydb/dragonfly/blob/4835b5debc955a60e85d1137b6da0531ec8cc380/.github/workflows/docker-weekly.yml#L12

for html publish action to work

File: tests/fakeredis/poetry.lock [hunk 2]
```diff
@@ -0,0 +1,595 @@
+# This file is automatically @generated by Poetry 1.8.3 and should not be changed by hand.
+
```
Comment (romange) [inline tests/fakeredis/poetry.lock:2]: do we have to submit it, or we should add it to .gitignore?

Comment (cunla) [reply tests/fakeredis/poetry.lock:2]: Generally speaking, it should be committed, so if anyone wants to run the tests locally, they will have the exact same versions as the workflow has.
(See also [official documentation](https://python-poetry.org/docs/basic-usage/#:~:text=You%20should%20commit%20the%20poetry,of%20dependencies%20(more%20below).))

File: .github/workflows/test-fakeredis.yml (hunk 1 above)
Comment (cunla) [reply .github/workflows/test-fakeredis.yml:4]: I divided the workflow to several jobs (See [here](https://github.com/dragonflydb/dragonfly/actions/runs/10477664538)).

- Running tests (with matrix so each set of tests can be run separately, giving more visibility into what is working well vs. what isn't).
- Publishing HTML.
- Publishing workflow summary.

In the `Publishing HTML` part - I added `id-token:write` permission, so the permission applies only on the relevant part of the workflow.

Comment (cunla) [reply .github/workflows/test-fakeredis.yml:4]: I will look into why it doesn't work here.

**Review by kostasrim — COMMENTED** [review]
:100:

File: .github/workflows/test-fakeredis.yml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        PYTHON_VERSION:
+          - "3.12"
+        DRAGONFLY_VERSION:
+          - "latest"
+        tests:
+          - "test_json"
+          #          - "test_mixins"
```
Comment (kostasrim) [inline .github/workflows/test-fakeredis.yml:28]: Hi @cunla , thank you so much for this contribution :pray: 

Is there a reason that this is commented out ?

Comment (cunla) [reply .github/workflows/test-fakeredis.yml:28]: Good catch! Forgot to uncomment it (this test is slower compared with the others and I wanted to test the github page creation).

File: .github/workflows/test-fakeredis.yml (hunk 1 above)
Comment (cunla) [reply .github/workflows/test-fakeredis.yml:4]: Added the token, should be fine now.

**Comment by romange** [comment]
Ok, @cunla  seems that the issue with github pages is specific with PRs coming from a foreign fork (security thingy):
https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token#changing-the-permissions-in-a-forked-repository

I would like to ask you to comment out the pull_request trigger as we do not want to run these checks on every PR just yet. 
we will have to fix the problems in Dtragonfly first, and we will have to change the workflow to build a local dragonfly - otherwise running on pull request does not make sense.

**Review by romange — DISMISSED** [review]

**Comment by romange** [comment]
seems that our precommit checks fail. Can you please install the checks as explained here:
https://github.com/dragonflydb/dragonfly/blob/main/CONTRIBUTING.md#before-you-make-your-changes

this will also reformat and make the necessary changes to your commit.  
see here for more details:
https://github.com/dragonflydb/dragonfly/actions/runs/10493048837/job/29072056658

**Comment by cunla** [comment]
Should be good now.

**Review by romange — APPROVED** [review]

**Pull request merged**